docker/freebsd-install.sh

```diff
@@ -6,35 +6,147 @@ set -euo pipefail
+    "pkg0.tuk.freebsd.org"
+    "pkg0.twn.freebsd.org"
+)
 
-setup_packagesite() {
-    apt-get update && apt-get install --assume-yes --no-install-recommends \
-        curl \
-        jq \
-        xz-utils
+# NOTE: out of convention, we use `url` for mirrors with the scheme,
+# and `mirror` for those without the scheme for consistent naming.
+freebsd_package_source() {
+    local url="${1}"
+    echo "${url}/FreeBSD:${FREEBSD_MAJOR}:${FREEBSD_ARCH}/quarterly"
+}
+
+freebsd_mirror_works() {
+    local mirror="${1}"
+    local scheme="${2}"
+    local pkg_source=
+
+    # meta.conf is a small file for quick confirmation the mirror works
+    pkg_source=$(freebsd_package_source "${scheme}://${mirror}")
+    local path="${pkg_source}/meta.conf"
+
+    timeout 20s curl --retry 3 -sSfL "${path}" >/dev/null 2>&1
+}
 
-    mkdir /opt/freebsd-packagesite
-    curl --retry 3 -sSfL "${PKG_SOURCE}/packagesite.txz" -O
-    tar -C /opt/freebsd-packagesite -xJf packagesite.txz
+_fetch_best_freebsd_mirror() {
+    # in case if the default mirror is down, we can use various known
+    # fallbacks, or at worst, SRV fallbacks to find the ideal package
+    # site. no individual mirror other than the default mirror is
+    # guaranteed to exist, so we use a tiered approach. only
+    # the default mirror supports https.
+    if freebsd_mirror_works "${FREEBSD_DEFAULT_MIRROR}" "https"; then
+        echo "https://${FREEBSD_DEFAULT_MIRROR}"
+        return 0
+    fi
+
+    # if we've gotten here, it could be a DNS issue, so using a DNS
+    # resolver to fetch SRV fallbacks may not work. let's first try
+    # a few previously tested mirrors and see if any work.
+    local mirror=
+    for mirror in "${FREEBSD_BACKUP_MIRRORS[@]}"; do
+        if freebsd_mirror_works "${mirror}" "http"; then
+            echo "http://${mirror}"
+            return 0
+        fi
+    done
+
+    local http_tcp_source=
+    local response=
+    local lines=
+    # shellcheck disable=SC2016
+    local regex='/\d+\s+\d+\s+\d+\s+(.*)\./; print $1'
+    for http_tcp_source in "${FREEBSD_HTTP_TCP_SOURCES[@]}"; do
+        # the output will have the following format, but we only want the
+        # target and ignore everything else:
+        #   $priority $port $weight $target.
+        #
+        # some output may not match, so we skip those lines, for example:
+        #   96.47.72.71
+        response=$(dig +short srv "${http_tcp_source}")
+        readarray -t lines <<< "${response}"
+        for line in "${lines[@]}"; do
+            mirror=$(echo "${line}" | perl -nle "${regex}")
+            if [[ -n "${mirror}" ]]; then
+                if freebsd_mirror_works "${mirror}" "http"; then
+                    echo "http://${mirror}"
+                    return 0
+                fi
+            fi
+        done
+    done
+
+    echo -e "\e[31merror:\e[0m could not find a working FreeBSD package mirror." 1>&2
+    exit 1
+}
+
+fetch_best_freebsd_mirror() {
+    set +e
```

Comment:
I've added wrappers for the functions since the flow control inside has multiple exit points, just to ensure code downstream exits on error.